mantidproject/mantid
# Reflectometry GUI restructuring 9: set defaults

~~***PR #25605 should be merged first***~~

**Description of work.**
Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR adds functionality to populate settings on the GUI with defaults taken from the reduction algorithm or parameters file.

- The changes to files in `API` and `Algorithms` are due to moving `BoostOptionalToAlgorithmProperty` to  `API` to make it available from the GUI.
- The changes to `ReflectometryHelper` to use the `AlgorithmManager` were necessary to use it from the GUI tests.
- Release notes have been added for all changes on the feature branch so far, so they include a lot more than just the content of this PR. The relevant part to this PR is that the defaults are now set automatically on the Experiment and Instrument tabs. I added the rest because this change makes more sense in the overall context.


**To test:**

- Set your default facility to `ISIS` and default instrument to `INTER`
- Open the `ISIS Reflectometry` interface
- Check that the populated values in the `Experiment` and `Instrument` tabs match the values in the `INTER_parameters.xml` file:

![image](https://user-images.githubusercontent.com/12895056/56351975-2f825400-61c6-11e9-89e5-edf0b8d0bd1c.png)

- On the `Runs` tab, change the instrument to `OFFSPEC` and check that the values have been updated to match the `OFFSPEC_Parameters.xml` file.

![image](https://user-images.githubusercontent.com/12895056/56352796-dddac900-61c7-11e9-8383-f89b8ad2a9ac.png)

- Note that at the moment if you toggle to the `Batch 1` tab on the left or add new batches, they will have the original instrument (INTER). We may change this in future but for now just check that the defaults still match the instrument for the batch.
- Change some values and then click `Restore Defaults`. Check that values from the parameters file are re-instated and any error highlighting is cleared. 
- Edit `INTER_Parameters.xml` in your local instrument directory (e.g. `~/.mantid/instrument` on Linux or `%APPDATA%/mantidproject/mantid/instrument` on Windows). Edit some values to make them invalid e.g. set `LambdaMax` to be less than `LambdaMin`. Restart mantid and open the interface. Check that an error is reported in the log and that the relevant inputs on the interface are unset.
  - Note that entering invalid types in the parameters file e.g. a string value when a numeric value is expected, is currently treated as unset rather than reporting an error; this is perhaps not ideal but is due to the way that `BoostOptionalToAlgorithmProperty` works and changing that is beyond the scope of this PR.

Fixes #23046 and Refs #23051

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
